src/blog/better-class-features.md

```diff
@@ -0,0 +1,332 @@
+LdaKeyedProperty <this>, [0]
+Star r2
+
+// Move the target into r0
+Mov <this>, r0
+
+// Use the %CreateDataProperty() runtime function to store the property keyed
+// by `#a` as the value of the property keyed by `b`, that is, `b = this.#a`
+CallRuntime [CreateDataProperty], r0-r2
+```
+
+Compare the class in the previous snippet to a class like this:
+
+```js
+class A {
+  constructor() {
+    this._a = 0;
+    this.b = this._a;
+  }
+}
+```
+
+Technically these two classes are not equivalent, even ignoring the difference in visibility between `this.#a` and `this._a` - as the specification mandates, initialization of the class fields does not trigger setters or `set` Proxy traps. So an approximation of the first class should use `Object.defineProperty()` instead of simple assignments to initialize the properties. In addition, it should throw if the private field already exists in the instance (in case the target being initialized is overridden in the base constructor to be another instance):
+
+```js
+class A {
+  constructor() {
+    // What the %AddPrivateField() call roughly translates to:
+    const _a = %PrivateSymbol('#a')
+    if (_a in this) {
+      throw TypeError('Cannot initialize #a twice on the same object');
+    }
+    Object.defineProperty(this, _a, {
+      writable: true,
+      configurable: false,
+      enumerable: false,
+      value: 0
+    });
+    // What the %CreateDataProperty() call roughly translates to:
+    Object.defineProperty(this, 'b', {
+      writable: true,
+      configurable: true,
+      enumerable: true,
+      value: this[_a]
+    });
+  }
+}
+```
+
+To implement the specified semantics before the proposal finalized, V8 used calls to runtime functions since they are more flexible. As shown in the bytecode above, the initialization of public fields was implemented with `%CreateDataProperty()` runtime calls, while the initialization of private fields was implemented with `%AddPrivateField()`. Since calling into the runtime incurs a significant overhead, the initialization of class fields was significantly slower compared to the assignment of ordinary object properties.
+
+In most use cases, however, the semantic differences are insignificant. It would be nice to have the performance of the optimized assignments of properties when the pattern of the class field initialization is predictable enough - so a more optimal implementation was created after the proposal finalized.
+
+### Optimizing private class fields and computed public class fields
+
+To speed up initialization of private class fields and computed public class fields, the implementation introduced a new machinery to plug into the [inline cache (IC) system](https://mathiasbynens.be/notes/shapes-ics) when handling these operations, including:
+
+- A new bytecode `DefineKeyedOwnProperty`. This gets emitted when generating code for the `ClassLiteral::Property` AST nodes representing class field initializers.
```

Comment:
```suggestion
- In the interpreter, a new bytecode `DefineKeyedOwnProperty`. This gets emitted when generating code for the `ClassLiteral::Property` AST nodes representing class field initializers.
```

Reply:
I am wondering if it's accurate to phrase it as an interpreter thing, as the bytecode is also compiled by the compilers. Perhaps `In the bytecode generator` would be more accurate?

Reply:
Fair point. `In the bytecode generator` sgtm.